flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java

```diff
@@ -1955,6 +1959,25 @@ private void abortPendingCheckpoint(
         }
     }
 
+    private void handleCheckpointException(
+            PendingCheckpoint pendingCheckpoint,
+            CheckpointProperties checkpointProperties,
+            CheckpointException exception,
+            ExecutionAttemptID executionAttemptID) {
+        if (checkpointProperties.isSavepoint()
+                && (checkpointProperties.isSynchronous() || pendingCheckpoint == null)) {
+            failureManager.handleSynchronousSavepointFailure(exception);
```

Comment:
I don't understand this condition. Why `pendingCheckpoint == null` and non-synchronous savepoints are falling into `handleSynchronousSavepointFailure(...)`?

Reply:
My idea is if we fail during the initialization phase it means that asynchronous part of the checkpoint/savepoint has not started yet so we can do the same logic as for synchronous savepoint

Reply:
async phase of checkpoint has unfortunately nothing to do with "synchronous savepoint". Synchronous savepoint is a bad internal name for a savepoint barrier that is part of "stop-with-savepoint" feature. In this context, non-synchronous savepoint are savepoints that are not followed by stopping of the job.